src/main/java/com/maxmind/db/Decoder.java

```diff
@@ -220,65 +207,257 @@ static int decodeInteger(ByteBuffer buffer, int base, int size) {
         return integer;
     }
 
-    private BigIntegerNode decodeBigInteger(int size) {
+    private BigInteger decodeBigInteger(int size) {
         byte[] bytes = this.getByteArray(size);
-        return new BigIntegerNode(new BigInteger(1, bytes));
+        return new BigInteger(1, bytes);
     }
 
-    private DoubleNode decodeDouble(int size) throws InvalidDatabaseException {
+    private double decodeDouble(int size) throws InvalidDatabaseException {
         if (size != 8) {
             throw new InvalidDatabaseException(
                     "The MaxMind DB file's data section contains bad data: "
                             + "invalid size of double.");
         }
-        return new DoubleNode(this.buffer.getDouble());
+        return this.buffer.getDouble();
     }
 
-    private FloatNode decodeFloat(int size) throws InvalidDatabaseException {
+    private float decodeFloat(int size) throws InvalidDatabaseException {
         if (size != 4) {
             throw new InvalidDatabaseException(
                     "The MaxMind DB file's data section contains bad data: "
                             + "invalid size of float.");
         }
-        return new FloatNode(this.buffer.getFloat());
+        return this.buffer.getFloat();
     }
 
-    private static BooleanNode decodeBoolean(int size)
+    private static boolean decodeBoolean(int size)
             throws InvalidDatabaseException {
         switch (size) {
             case 0:
-                return BooleanNode.FALSE;
+                return false;
             case 1:
-                return BooleanNode.TRUE;
+                return true;
             default:
                 throw new InvalidDatabaseException(
                         "The MaxMind DB file's data section contains bad data: "
                                 + "invalid size of boolean.");
         }
     }
 
-    private JsonNode decodeArray(int size) throws IOException {
+    private <T> Object decodeArray(int size, Class<T> cls)
+            throws IOException,
+                   InstantiationException,
+                   IllegalAccessException,
+                   InvocationTargetException,
+                   DeserializationException {
+        // If we're decoding into a Map, the second level is an Object.class.
+        // See Object.class in decodeMapIntoMap().
+        if (cls.equals(List.class) || cls.equals(Object.class)) {
+            ArrayList<Object> array = new ArrayList<>(size);
+
+            for (int i = 0; i < size; i++) {
+                Object e = this.decode(Object.class);
```

Comment:
Will this work for the subdivision list?

Reply:
Hmm apparently not :(. Not quite sure how to get that working currently. I'll poke at it.

Reply:
Not really sure it is possible so far though given type erasure.

Reply:
Jackson Databind is obviously doing it somehow. :shrug: